<a href="https://colab.research.google.com/github/jeongy72/py25/blob/master/notebook/mid_exam_py25.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## 파이썬 활용 중간고사 풀이

### 문제1. 다음 식의 연산 결과는 ?

In [0]:
 a = 5   
 b = 3    
 print(a**b)    

125


#### 문제풀이:
** 연산자는 a의 b제곱을 나타내는 연산자 이므로 125라는 결과가 나옴.

### 문제2. 다음 문자열 a에 대하여 슬라이싱을 한 결과는 ?

In [0]:
a = "Life is too short,You need Python!"
a[-16:-8]

'You need'

#### 문제풀이:
슬라이싱은 복잡하고 긴 문자열에서 원하는 부분만 뽑아낼 수 있는 기법이다.
위 소스 코드에서 a[-16:-8]이 뜻하는 것은 a[-16]에서 [-9]까지를 말한다.마이너스가 붙었으니 뒤에서 부터 세면 'You need'라는 결과가 나옴.

### 문제3. 다음 list a의 인덱싱 결과는 ?

In [0]:
a = [1,2,(3,4),['a','b',['My','Life','Good']]]
a[3][-1][0]

'My'

#### 문제풀이:
인덱싱은 리스트에서 특정한 부분을 가르키는 것이다.
파이썬은 0부터 시작한다. 그래서 세 번째 리스트인 ['My','Life','Good']에서 [0]인 'My'인 결과가 나옴.

### 문제4. 다음 tuple t1의 슬라이싱 결과는 ?

In [0]:
t1 = (0,1,"s","t","u")
t1[2:-2]

('s',)

#### 문제풀이:
[2:-2]는 2부터 시작해서 -3까지니까 세보면 's'가 나온다. 하지만 튜플은 한 가지 요소만 있을 때는 반드시 ,가 붙기 때문에 ('s',)라는 결과가 나옴.

### 문제5. 다음 bool값 출력 중 정확한 것은 ?

In [0]:
bool([0,1,2]), bool([0]), bool((0)), bool((0,))

(True, True, False, True)

#### 문제풀이:
bool자료형은 True, False 값만 가질 수 있다. 튜플 구조는 요소 하나만 있어도 반드시 ,가 있어야 하기에 세번 째 bool만 False이고 나머지는 True로 결과값이 나옴.

### 문제6. 다음 조건문을 1줄 코드로 변경하시오.

In [5]:
score = 80
if score >= 60:
    message = "success"
else:
     message = "failure"
print(score,message)

80 success


In [6]:
score = 80
meesage = "success" if score >= 60 else "failure"
print(score,message)

80 success


#### 문제풀이:
위의 보기 코드는 조건문을 이용하여 만든 것이다. 한 줄로 코드를 변경하려면 위와 같이 단축코드를 이용하면 결과값이 같게 나옴. 

### 문제7. 다음 코드의 실행 결과는 ?

In [9]:
games = ["A","B","C"]

for cnt in range(len(games)):
    print(cnt,games[cnt], end = ',')


0 A,1 B,2 C,

#### 문제풀이:
for문을 이용한 코드이다. games라는 변수의 길이를 games안에 들어있는 각각의 변수 앞에 출력 시키고 한 항목 사이에는 그냥 띄우고, 항목과 항목 사이에는 줄바꿈없이 end 속성으로 end에 ,가 출력되도록 하였다.그래서 위와 같이 출력이 됨.

### 문제8. 다음의 결과가 출력되지 않는  것은?
' I ate 10 apples, so i was sick for three days.

In [10]:
number = 10
day = "three"
"I ate {0} apples, so i was sick for {1} days.".format(10,day)

'I ate 10 apples, so i was sick for three days.'

#### 문제풀이:
format함수를 사용한 포매팅으로 숫자를 바로 대입한 것이다. 그래서 보기와 같이 출력됨.

In [12]:
number = 10
day = "three"
f"I ate {number} apples, so i was sick for {'three'} days."

'I ate 10 apples, so i was sick for three days.'

#### 문제풀이:
f문자열 포매팅을 사용한 코드로 문자열 앞에 f접두사를 붙이면 f문자열 포매팅 기능을 사용할 수 있다. 그래서 보기와 같이 출력됨.

In [13]:
number = 10
day = "three"
f"I ate {number} apples, so i was sick for {day} days."

'I ate 10 apples, so i was sick for three days.'

#### 문제풀이:
이것 또한 f문자열 포매팅을 사용한 코드이다. 그래서 보기와 같이 출력됨.

In [14]:
number = 10
day = "three"
"I ate {0} apples, so i was sick for {1} days.".format(number=10,day)

SyntaxError: ignored

#### 문제풀이:
format함수를 이용한 포매팅 코드이다. 인덱스와 이름을 혼용하여 넣을려면 {}안에도 맞게 =숫자가 있을때는 이름으로 , 이름이 있으면 인덱스가 있어야 한다. 예를 들어{1}이면 format(day)이여야 하고, {number}이면  format(number=10)이여야 한다. 그래서 #### 문제풀이:
format함수를 이용한 포매팅 코드이다. 인덱스와 이름을 혼용하여 넣을려면 {}안에도 맞게 =숫자가 있을때는 이름으로 , 이름이 있으면 인덱스가 있어야 한다. 예를 들어{1}이면 format(day)이여야 하고, {number}이면  format(number=10)이여야 한다. 그래서 위치 인수는 키워드 인수를 따르기 때문에 에러가 발생함.

### 문제9. 라스트 내포를 사용해 1부터 1000이하의 자연수 중 5의 배수의 합을 구하는 한 줄 코드를 작성하시오.

In [19]:
sum([num for num in range(1,1001) if num % 5 ==0])


100500

#### 문제풀이:
sum은 숫자를 더하는 함수이고, 거기에 for문을 이용하여 5로 나누었을 때 0으로 딱 떨어지는 숫자만 더하도록 코드를 만들었다. 그래서 위와 같은 결과가 나옴.

### 문제10. 4개의 과목의 평균 성적을 구하시오.


In [23]:
subjects =  ['python','english','math','science']
scores = [88,77,66,95]
total= 0

for subject, score in zip(subjects, scores):
    total +=  score
    print(subject, score)

print("평균 점수 = {0}".format(total/len(subjects)))

python 88
english 77
math 66
science 95
평균 점수 = 81.5


#### 문제풀이:
socre에 따로 position을 하지 않았으니 함수 zip를 썼을 것이고, 평균을 구해야 하니까 format함수를 이용하여 전체 더한 점수에서 과목 길이 만큼을 나눈 것이다. 그래서[a]= zip이고,[b] = .format(total/len(subjects)) 가 됨.
